src/neural/cuda/layers.cc

```diff
@@ -900,6 +901,251 @@ FusedWinogradConvSELayer<DataType>::~FusedWinogradConvSELayer() {
+                              cudaMemcpyHostToDevice));
+  copyTypeConverted((DataType*)b1_, (float*)scratch, (int)num_biases1);
+
+  ReportCUDAErrors(cudaMemcpy(scratch, b2, num_biases2 * sizeof(float),
+                              cudaMemcpyHostToDevice));
+  copyTypeConverted((DataType*)b2_, (float*)scratch, (int)num_biases2);
+}
+
+template <>
+void ResidualBlock<half>::cublasRowMajorMatrixMul(
+    const half* A, const half* B, half* Out, int M, int N, int K, int batchSize,
+    cublasHandle_t cublas) {
+  // Need to initialize 1.0 and 0.0 as hexadecimal for fp16 because typecasting
+  // float to half type doesn't work before CUDA 10.0
+  __half_raw one_h{0x3C00};
+  __half_raw zero_h{0};
+  half halfOne = one_h;
+  half halfZero = zero_h;
+
+  // dimensions of matrix A = M x K
+  // dimensions of matrix B = K x N
+  // dimensions of output   = M x N
+
+  // cublas supports only col major output
+  // to multiply row major matrices, use the trick below
+  ReportCUBLASErrors(cublasGemmStridedBatchedEx(
+      cublas, CUBLAS_OP_N, CUBLAS_OP_N, N, M, K, &halfOne, B, CUDA_R_16F, N,
+      N * K, A, CUDA_R_16F, K, K * M, &halfZero, Out, CUDA_R_16F, N, N * M,
+      batchSize, CUDA_R_16F, CUBLAS_GEMM_DEFAULT));
+}
+
+template <>
+void ResidualBlock<float>::cublasRowMajorMatrixMul(
+    const float* A, const float* B, float* Out, int M, int N, int K,
+    int batchSize, cublasHandle_t cublas) {
+  float floatOne = 1.0f;
+  float floatZero = 0.0f;
+  if (use_gemm_ex_)
+    ReportCUBLASErrors(cublasGemmStridedBatchedEx(
+        cublas, CUBLAS_OP_N, CUBLAS_OP_N, N, M, K, &floatOne, B, CUDA_R_32F, N,
+        N * K, A, CUDA_R_32F, K, K * M, &floatZero, Out, CUDA_R_32F, N, N * M,
+        batchSize, CUDA_R_32F, CUBLAS_GEMM_DEFAULT));
+  else
+    ReportCUBLASErrors(cublasSgemmStridedBatched(
+        cublas, CUBLAS_OP_N, CUBLAS_OP_N, N, M, K, &floatOne, B, N, N * K, A, K,
+        K * M, &floatZero, Out, N, N * M, batchSize));
+}
+
+template <typename DataType>
+void ResidualBlock<DataType>::Eval(
+    int N, DataType* output, const DataType* input, const DataType* /*input2*/,
+    void* scratch, size_t scratch_size, cudnnHandle_t /*cudnn*/,
+    cublasHandle_t cublas) {
+
+  // normally:
+  // - "output" initially contains the transformed input, 
+  //    and after this layer, it contains the transformed input for next layer
+  // - "input" contains the original/untransformed input
+  // special cases:
+  //   - for first_block_, input is real input (untransformed)
+  //   - for last_block_, output is the final output of this block (untransformed)
+
+  // Split the scratch space into two parts - use first part for holding
+  // transformed input and second part for transformed output.
+  DataType* transformed_input = (DataType*)scratch;
+  DataType* transformed_output =
+      transformed_input + scratch_size / (2 * sizeof(DataType));
+
+  if (first_block_) {
+    InputTransform<DataType>(N, C, transformed_input, input);
```

Comment:
This should be `InputTransform<DataType>(N, c_input_, transformed_input, input);`

Reply:
Thanks. Will fix it. It works because right now the network has c_input_ == C for all the convolutions in residual tower, but it's better to not rely on this assumption.